### ISSUE TITLE
Trying to deploy gives me: TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received undefined

### ISSUE THREAD

**Issue by harrisrobin** [firebase/firebase-tools#3300]
### [REQUIRED] Environment info

**firebase-tools:** 9.10.0

**Platform:** macOS

### [REQUIRED] Test case

When I run 
```sh
firebase deploy --only functions
```

I get the following (firebase-debug.log):

```
[debug] [2021-04-22T05:25:42.747Z] ----------------------------------------------------------------------
[debug] [2021-04-22T05:25:42.749Z] Command:       /Users/OMITTED/.volta/tools/image/node/14.16.1/bin/node /Users/OMITTED/.volta/tools/image/packages/firebase-tools/bin/firebase deploy
[debug] [2021-04-22T05:25:42.749Z] CLI Version:   9.10.0
[debug] [2021-04-22T05:25:42.749Z] Platform:      darwin
[debug] [2021-04-22T05:25:42.749Z] Node Version:  v14.16.1
[debug] [2021-04-22T05:25:42.750Z] Time:          Wed Apr 21 2021 22:25:42 GMT-0700 (Pacific Daylight Time)
[debug] [2021-04-22T05:25:42.750Z] ----------------------------------------------------------------------
[debug] 
[debug] [2021-04-22T05:25:42.755Z] > command requires scopes: ["email","openid","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","https://www.googleapis.com/auth/cloud-platform"]
[debug] [2021-04-22T05:25:42.756Z] > authorizing via signed-in user (EMAIL_OMITTED)
[debug] [2021-04-22T05:25:42.756Z] [iam] checking project PROJECT_NAME-dev for permissions ["cloudfunctions.functions.create","cloudfunctions.functions.delete","cloudfunctions.functions.get","cloudfunctions.functions.list","cloudfunctions.functions.update","cloudfunctions.operations.get","datastore.indexes.create","datastore.indexes.delete","datastore.indexes.list","datastore.indexes.update","firebase.projects.get"]
[debug] [2021-04-22T05:25:42.757Z] >>> HTTP REQUEST POST https://cloudresourcemanager.googleapis.com/v1/projects/PROJECT_NAME-dev:testIamPermissions  
 {"permissions":["cloudfunctions.functions.create","cloudfunctions.functions.delete","cloudfunctions.functions.get","cloudfunctions.functions.list","cloudfunctions.functions.update","cloudfunctions.operations.get","datastore.indexes.create","datastore.indexes.delete","datastore.indexes.list","datastore.indexes.update","firebase.projects.get"]}
[debug] [2021-04-22T05:25:42.930Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Thu, 22 Apr 2021 05:25:42 GMT","server":"ESF","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","server-timing":"gfet4t7; dur=45","alt-svc":"h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
[debug] [2021-04-22T05:25:42.932Z] >>> HTTP REQUEST POST https://iam.googleapis.com/v1/projects/PROJECT_NAME-dev/serviceAccounts/PROJECT_NAME-dev@appspot.gserviceaccount.com:testIamPermissions  
 {"permissions":["iam.serviceAccounts.actAs"]}
[debug] [2021-04-22T05:25:43.091Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Thu, 22 Apr 2021 05:25:43 GMT","server":"ESF","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","alt-svc":"h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
[debug] [2021-04-22T05:25:43.231Z] TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received undefined
    at validateString (internal/validators.js:124:11)
    at Object.join (path.js:1039:7)
    at Object.checkFunctionsSDKVersion [as fn] (/Users/OMITTED/.volta/tools/image/packages/firebase-tools/lib/node_modules/firebase-tools/lib/checkFirebaseSDKVersion.js:34:28)
    at /Users/OMITTED/.volta/tools/image/packages/firebase-tools/lib/node_modules/firebase-tools/lib/command.js:169:30
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
[error] 
[error] Error: An unexpected error has occurred.
```

I'm not sure what's wrong. I'm an owner in the firebase project, I have verified that I am using the correct projectId and I setup the functions repo using `firebase init functions`. 

Any help is appreciated.

### [REQUIRED] Steps to reproduce

Run
```sh
firebase deploy --only functions
```

### [REQUIRED] Expected behavior

For the deployment to work

### [REQUIRED] Actual behavior

I get the error above.

**Comment by samtstern** [comment]
@harrisrobin could you show me your `firebase.json` file?

**Comment by harrisrobin** [comment]
> @harrisrobin could you show me your `firebase.json` file?

@samtstern thank you for taking a look. Here's my `firebase.json` file
```
{
  "functions": {
    "predeploy": [
      "npm --prefix \"$RESOURCE_DIR\" run lint",
      "npm --prefix \"$RESOURCE_DIR\" run build"
    ]
  },
  "firestore": {
    "rules": "firestore.rules",
    "indexes": "firestore.indexes.json"
  },
  "emulators": {
    "firestore": {
      "port": "8080"
    }
  }
}
```

**Comment by samtstern** [comment]
I believe this will be fixed if you just add a `source` directory to your `functions` config:

```js
  "functions": {
   // ADD THIS
   "source": "functions",
    "predeploy": [
      "npm --prefix \"$RESOURCE_DIR\" run lint",
      "npm --prefix \"$RESOURCE_DIR\" run build"
    ]
  },
  ```
  
  That said the error could be a lot better! So I'll send a PR to fix that. But let me know if that works for you.

**Comment by harrisrobin** [comment]
> ```js
>    "source": "functions",
> ```

@samtstern  That worked 🎉 

Thank you good sir. 

FYI, for anyone else I also had to remove the `predeploy` field because `npm` could not find `$RESOURCE_DIR` for some reason.

Going to close this.

**Comment by samtstern** [comment]
Nice glad it worked!  I want to open this real quick to confirm on our side that this is the behavior we want.

**Comment by samtstern** [comment]
@joehan just checking if we should relax this at all ... is it our stance that if you supply any config in `functions` then we no longer assume a default value for `functions.source`?

**Comment by harrisrobin** [comment]
@samtstern in that case, i'd just like to give you more context.

My folder directory is:

```
root
---src
------index.ts (functions are here)
```

I do not have a folder called "functions". 

What i did with source is actually make it the following:
```json
{
  "functions": {
    "source": "."
  },
  "firestore": {
    "rules": "firestore.rules",
    "indexes": "firestore.indexes.json"
  },
  "emulators": {
    "firestore": {
      "port": "8080"
    }
  }
}
```

and not "functions" although i understood what you meant. Hope this helps.

**Comment by samtstern** [comment]
@harrisrobin ahhh ok that does make a difference! Since we have code to look for `functions/package.json` if we don't detect a `source` value but we can't guess if you've chosen something else. 

So I think this now comes down to just an issue of a bad error message.

**Comment by harrisrobin** [comment]
> @harrisrobin ahhh ok that does make a difference! Since we have code to look for `functions/package.json` if we don't detect a `source` value but we can't guess if you've chosen something else.
> 
> So I think this now comes down to just an issue of a bad error message.

yeah that makes sense and agreed. If the error had some indication that I had to specify a `source` it would have helped. Thanks again for looking into this! Hopefully this issue helps someone else as well.

**Comment by joehan** [comment]
Thanks for reporting this @harrisrobin , this is definitely something we can improve the experience on. 
I opened a PR that should catch this case, as well as the case when functions.source is defined but incorrect.

@samtstern, the behavior here is a bit weird: we only set ./functions as a default, if we see a package.json there (see https://github.com/firebase/firebase-tools/blob/master/src/config.js#L55). Otherwise, we just leave it empty (and then break).
